File: go/jsonrpc3/http_server.go
package jsonrpc3

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"
	"time"
)

const (
	// DefaultSessionTTL is the default time-to-live for HTTP sessions with refs
	DefaultSessionTTL = 5 * time.Minute

	// StatelessSessionTTL is the time-to-live for sessions without object references
	// These are cleaned up more aggressively to optimize memory usage
	StatelessSessionTTL = 30 * time.Second

	// SessionCleanupInterval is how often we check for expired sessions
	SessionCleanupInterval = 10 * time.Second
)

// HTTPHandler is an HTTP handler that serves JSON-RPC 3.0 requests.
// It reads requests from HTTP POST bodies, processes them using a Handler,
// and writes responses back to the HTTP response.
// Sessions are tracked via RPC-Session-Id header and automatically expire.
type HTTPHandler struct {
	rootObject   Object
	mimeTypes    []string
	version      string
	sessions     map[string]*sessionEntry
	sessionMutex sync.RWMutex
	sessionTTL   time.Duration
	stopCleanup  chan struct{}
}

// sessionEntry tracks a session, its handler, and last access time
type sessionEntry struct {
	session      *Session
	handler      *Handler
	lastAccessed time.Time
	mutex        sync.Mutex
}

// NewHTTPHandler creates a new HTTP handler for JSON-RPC 3.0.
// The rootObject handles top-level method calls.
// Sessions can span multiple requests via the RPC-Session-Id header.
// Supports all three encoding formats: JSON, CBOR, and compact CBOR.
func NewHTTPHandler(rootObject Object) *HTTPHandler {
	h := &HTTPHandler{
		rootObject: rootObject,
		mimeTypes: []string{
			MimeTypeJSON,
			MimeTypeCBOR,
			MimeTypeCBORCompact,
		},
		version:     Version30,
		sessions:    make(map[string]*sessionEntry),
		sessionTTL:  DefaultSessionTTL,
		stopCleanup: make(chan struct{}),
	}

	// Start background cleanup goroutine
	go h.cleanupLoop()

	return h
}

// SetSessionTTL sets the session time-to-live duration.
// Sessions that haven't been accessed for this duration will be expired.
func (h *HTTPHandler) SetSessionTTL(ttl time.Duration) {
	h.sessionMutex.Lock()
	defer h.sessionMutex.Unlock()
	h.sessionTTL = ttl
}

// Close stops the background cleanup goroutine.
// Call this when shutting down the handler.
func (h *HTTPHandler) Close() {
	close(h.stopCleanup)
}

// ServeHTTP implements http.Handler.
// It processes JSON-RPC requests and returns responses.
// Sessions are reused if RPC-Session-Id header is present, otherwise a new session is created and stored.
// Sessions without object references are cleaned up more aggressively to optimize memory usage.
// DELETE method with RPC-Session-Id header deletes the session and returns 204.
func (h *HTTPHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	// Handle DELETE method for session deletion
	if r.Method == http.MethodDelete {
		sessionID := r.Header.Get("RPC-Session-Id")
		if sessionID != "" {
			h.sessionMutex.Lock()
			entry, exists := h.sessions[sessionID]
			if exists {
				// Dispose all refs before deleting session
				entry.session.DisposeAll()
				delete(h.sessions, sessionID)
			}
			h.sessionMutex.Unlock()
		}
		w.WriteHeader(http.StatusNoContent)
		return
	}

	// Only accept POST requests for RPC
	if r.Method != http.MethodPost {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
		return
	}

	// Get content type from request
	contentType := r.Header.Get("Content-Type")
	if contentType == "" {
		contentType = MimeTypeJSON
	}

	// Read request body
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, "Failed to read request body", http.StatusBadRequest)
		return
	}
	defer r.Body.Close()

	// Get or create session based on RPC-Session-Id header
	clientSessionID := r.Header.Get("RPC-Session-Id")
	session, handler, sessionID, wasExisting := h.getOrCreateSession(clientSessionID)

	// Decode request
	codec := GetCodec(contentType)
	msgSet, err := codec.UnmarshalMessages(body)
	if err != nil {
		// Send parse error response
		errResp := NewErrorResponse(nil, NewParseError(err.Error()), h.version)
		h.writeResponse(w, errResp, contentType, codec)
		return
	}

	// Check if this was originally a batch (JSON/CBOR array)
	var req *Request
	var batch Batch
	var isBatch bool
	if !msgSet.IsBatch {
		req, err = msgSet.ToRequest()
		if err != nil {
			errResp := NewErrorResponse(nil, NewParseError(err.Error()), h.version)
			h.writeResponse(w, errResp, contentType, codec)
			return
		}
	} else {
		batch, err = msgSet.ToBatch()
		isBatch = true
		if err != nil {
			errResp := NewErrorResponse(nil, NewParseError(err.Error()), h.version)
			h.writeResponse(w, errResp, contentType, codec)
			return
		}
	}

	// Check if request contains client references for SSE mode
	// Only support SSE for single requests (not batch) for now
	useSSE := false
	if !isBatch && req != nil && scanForClientRefs(req.Params) {
		useSSE = true
	}

	// Handle SSE mode if client refs detected
	if useSSE {
		h.handleSSERequest(w, r, codec, req, handler, contentType, sessionID, wasExisting, session)
		return
	}

	// Process request (non-SSE mode)
	var batchResponses BatchResponse
	var resp *Response
	var batchData []byte
	var hasResponse bool

	if isBatch {
		// Handle batch request
		batchResponses = handler.HandleBatch(batch)
		hasResponse = len(batchResponses) > 0

		if hasResponse {
			var err error
			msgSet := batchResponses.ToMessageSet()
			batchData, err = codec.MarshalMessages(msgSet)
			if err != nil {
				http.Error(w, "Failed to encode response", http.StatusInternalServerError)
				return
			}
		}
	} else {
		// Handle single request
		resp = handler.HandleRequest(req)
		hasResponse = resp != nil
	}

	// Determine if session should be stored
	// Store if: 1) existing session, 2) protocol method call ($rpc), 3) has local refs
	shouldStore := wasExisting
	isProtocolMethod := !isBatch && req != nil && (req.Ref == "$rpc" || strings.HasPrefix(req.Method, "$rpc."))
	hasRefs := false

	if !wasExisting {
		// Check if session has refs after processing
		hasRefs = len(session.ListLocalRefs()) > 0

		// Store if it's a protocol method call or has refs
		if isProtocolMethod || hasRefs {
			h.storeSession(sessionID, session, handler)
			shouldStore = true
		}
	}

	// Set session ID header only if session is stored
	if shouldStore {
		w.Header().Set("RPC-Session-Id", sessionID)
	}

	// Write response
	if isBatch {
		if hasResponse {
			h.writeData(w, batchData, contentType)
		} else {
			// All notifications - return 204 No Content
			w.WriteHeader(http.StatusNoContent)
		}
	} else {
		if hasResponse {
			h.writeResponse(w, resp, contentType, codec)
		} else {
			// Notification - return 204 No Content
			w.WriteHeader(http.StatusNoContent)
		}
	}
}

// writeResponse encodes and writes a single response.
func (h *HTTPHandler) writeResponse(w http.ResponseWriter, resp *Response, contentType string, codec Codec) {
	msgSet := resp.ToMessageSet()
	data, err := codec.MarshalMessages(msgSet)
	if err != nil {
		http.Error(w, "Failed to encode response", http.StatusInternalServerError)
		return
	}
	h.writeData(w, data, contentType)
}

// writeData writes encoded data with appropriate headers.
func (h *HTTPHandler) writeData(w http.ResponseWriter, data []byte, contentType string) {
	w.Header().Set("Content-Type", contentType)
	w.WriteHeader(http.StatusOK)
	w.Write(data)
}

// getOrCreateSession retrieves an existing session or creates a new one.
// Returns the session, handler, session ID, and whether it was an existing session.
// New sessions are NOT automatically stored - call storeSession after processing.
func (h *HTTPHandler) getOrCreateSession(sessionID string) (*Session, *Handler, string, bool) {
	now := time.Now()

	// Try to get existing session
	if sessionID != "" {
		h.sessionMutex.RLock()
		entry, exists := h.sessions[sessionID]
		h.sessionMutex.RUnlock()

		if exists {
			// Lock the session entry to update last accessed time
			entry.mutex.Lock()
			entry.lastAccessed = now
			entry.mutex.Unlock()
			return entry.session, entry.handler, sessionID, true
		}
	}

	// Create new session and handler (but don't store yet)
	session := NewSession()
	newSessionID := session.ID()
	handler := NewHandler(session, h.rootObject, h.mimeTypes)
	handler.SetVersion(h.version)

	return session, handler, newSessionID, false
}

// storeSession stores a session for reuse across requests.
// Sessions without local refs will be cleaned up more aggressively by the cleanup loop.
func (h *HTTPHandler) storeSession(sessionID string, session *Session, handler *Handler) {
	now := time.Now()
	entry := &sessionEntry{
		session:      session,
		handler:      handler,
		lastAccessed: now,
	}

	h.sessionMutex.Lock()
	h.sessions[sessionID] = entry
	h.sessionMutex.Unlock()
}

// cleanupLoop periodically removes expired sessions
func (h *HTTPHandler) cleanupLoop() {
	ticker := time.NewTicker(SessionCleanupInterval)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			h.cleanupExpiredSessions()
		case <-h.stopCleanup:
			return
		}
	}
}

// cleanupExpiredSessions removes sessions that haven't been accessed recently.
// Sessions without object references are cleaned up more aggressively.
func (h *HTTPHandler) cleanupExpiredSessions() {
	now := time.Now()

	h.sessionMutex.Lock()
	defer h.sessionMutex.Unlock()

	// Find expired sessions
	var toDelete []string
	for id, entry := range h.sessions {
		entry.mutex.Lock()
		lastAccessed := entry.lastAccessed
		entry.mutex.Unlock()

		// Determine TTL based on whether session has refs
		refs := entry.session.ListLocalRefs()
		ttl := h.sessionTTL
		if len(refs) == 0 {
			// Sessions without refs expire faster - use the minimum of configured TTL or StatelessSessionTTL
			if StatelessSessionTTL < ttl {
				ttl = StatelessSessionTTL
			}
		}

		if now.Sub(lastAccessed) > ttl {
			toDelete = append(toDelete, id)
		}
	}

	// Delete expired sessions and dispose their refs
	for _, id := range toDelete {
		if entry, exists := h.sessions[id]; exists {
			entry.session.DisposeAll()
		}
		delete(h.sessions, id)
	}
}

// scanForClientRefs recursively scans params for Reference structures.
// Returns true if any client references are found ({"$ref": "..."}).
func scanForClientRefs(params RawMessage) bool {
	if len(params) == 0 {
		return false
	}

	// Try to decode as generic structure
	var data any
	if err := json.Unmarshal(params, &data); err != nil {
		return false
	}

	return hasClientRefsInValue(data)
}

// hasClientRefsInValue recursively checks if a value contains Reference structures.
func hasClientRefsInValue(value any) bool {
	switch v := value.(type) {
	case map[string]any:
		// Check if this is a Reference
		if refStr, ok := v["$ref"].(string); ok && refStr != "" && len(v) == 1 {
			return true
		}
		// Recurse into map values
		for _, val := range v {
			if hasClientRefsInValue(val) {
				return true
			}
		}
	case []any:
		// Recurse into array elements
		for _, val := range v {
			if hasClientRefsInValue(val) {
				return true
			}
		}
	}
	return false
}

// writeSSEEvent writes a single SSE event in the format:
// data: <json>\n\n
func writeSSEEvent(w http.ResponseWriter, flusher http.Flusher, data []byte) error {
	if _, err := fmt.Fprintf(w, "data: %s\n\n", data); err != nil {
		return err
	}
	if flusher != nil {
		flusher.Flush()
	}
	return nil
}

// handleSSERequest handles a request that contains client references using SSE.
// It sends the response as an SSE stream, allowing for notifications to be sent
// to client callback objects during request processing.
func (h *HTTPHandler) handleSSERequest(
	w http.ResponseWriter,
	r *http.Request,
	codec Codec,
	req *Request,
	handler *Handler,
	contentType string,
	sessionID string,
	wasExisting bool,
	session *Session,
) {
	// Set SSE headers
	w.Header().Set("Content-Type", "text/event-stream")
	w.Header().Set("Cache-Control", "no-cache")
	w.Header().Set("Connection", "keep-alive")

	// Always store session for SSE mode (client refs detected)
	if !wasExisting {
		h.sessionMutex.Lock()
		h.sessions[sessionID] = &sessionEntry{
			session:      session,
			handler:      handler,
			lastAccessed: time.Now(),
		}
		h.sessionMutex.Unlock()
	}

	// Set session ID header
	w.Header().Set("RPC-Session-Id", sessionID)

	// Get flusher for streaming
	flusher, ok := w.(http.Flusher)
	if !ok {
		http.Error(w, "Streaming not supported", http.StatusInternalServerError)
		return
	}

	// Write headers immediately
	w.WriteHeader(http.StatusOK)
	flusher.Flush()

	// Process request
	resp := handler.HandleRequest(req)

	// Encode and send result as SSE event
	if resp != nil {
		msgSet := resp.ToMessageSet()
		respData, err := codec.MarshalMessages(msgSet)
		if err != nil {
			// Can't send error response at this point, connection is already open
			return
		}

		if err := writeSSEEvent(w, flusher, respData); err != nil {
			// Connection error, nothing we can do
			return
		}
	}

	// Connection closes when function returns
}

File: ts/src/http-server.ts
/**
 * HTTP server transport using Bun.serve
 * Automatically upgrades to WebSocket when upgrade headers are detected
 */

import type { Server, ServerWebSocket } from "bun";
import { Session, type RpcObject } from "./session.ts";
import { Handler } from "./handler.ts";
import { NoOpCaller } from "./types.ts";
import { getCodec, MimeTypeJSON } from "./encoding.ts";
import { toRequest, toBatch, batchResponseToMessageSet } from "./types.ts";
import { parseError, invalidRequestError } from "./error.ts";

export interface HttpServerOptions {
  port?: number;
  hostname?: string;
  mimeType?: string;
}

/**
 * HTTP Server for JSON-RPC 3.0
 */
export class HttpServer {
  private server?: Server;
  private codec;
  private handler: Handler;

  constructor(
    rootObject: RpcObject,
    private options: HttpServerOptions = {},
  ) {
    const mimeType = options.mimeType ?? MimeTypeJSON;
    this.codec = getCodec(mimeType);
    // HTTP requests don't support callbacks, so use NoOpCaller
    const session = new Session();
    this.handler = new Handler(session, rootObject, new NoOpCaller(), [mimeType]);
  }

  /**
   * Start the server
   */
  async start(): Promise<void> {
    const port = this.options.port ?? 3000;
    const hostname = this.options.hostname ?? "localhost";

    this.server = Bun.serve({
      port,
      hostname,
      fetch: async (req, server) => this.handleHttp(req, server),
      websocket: {
        open: (ws) => this.handleWebSocketOpen(ws),
        message: (ws, message) => this.handleWebSocketMessage(ws, message),
        close: (ws) => this.handleWebSocketClose(ws),
      },
    });

    console.log(`JSON-RPC server listening on http://${hostname}:${port}`);
  }

  /**
   * Stop the server
   */
  async stop(): Promise<void> {
    if (this.server) {
      this.server.stop();
      this.server = undefined;
    }
  }

  /**
   * Get server URL
   */
  url(): string | undefined {
    if (!this.server) return undefined;
    return `http://${this.server.hostname}:${this.server.port}`;
  }

  /**
   * Handle HTTP request
   */
  private async handleHttp(req: Request, server: Server): Promise<Response> {
    // Check for WebSocket upgrade
    if (this.isWebSocketUpgrade(req)) {
      // Upgrade to WebSocket
      if (server.upgrade(req)) {
        return undefined!; // Connection was upgraded
      }
      return new Response("WebSocket upgrade failed", { status: 500 });
    }

    // Only accept POST for JSON-RPC
    if (req.method !== "POST") {
      return new Response("Method not allowed", { status: 405 });
    }

    try {
      // Read body
      const bodyBytes = new Uint8Array(await req.arrayBuffer());

      // Decode messages
      let msgSet;
      try {
        msgSet = this.codec.unmarshalMessages(bodyBytes);
      } catch (error) {
        // Parse error
        const err = parseError(error instanceof Error ? error.message : String(error));
        const errorResp = {
          jsonrpc: "3.0",
          error: err.toObject(),
          id: null,
        };
        const respBytes = this.codec.marshal(errorResp);
        return new Response(respBytes, {
          status: 200,
          headers: { "Content-Type": this.codec.mimeType() },
        });
      }

      // Handle based on batch or single
      let responseMsgSet;
      if (msgSet.isBatch) {
        // Batch request
        const batch = toBatch(msgSet);
        const batchResp = await this.handler.handleBatch(batch);
        responseMsgSet = batchResponseToMessageSet(batchResp);
      } else {
        // Single request
        const request = toRequest(msgSet);
        const response = await this.handler.handleRequest(request);
        responseMsgSet = {
          messages: [response],
          isBatch: false,
        };
      }

      // Encode response
      const respBytes = this.codec.marshalMessages(responseMsgSet);

      return new Response(respBytes, {
        status: 200,
        headers: { "Content-Type": this.codec.mimeType() },
      });
    } catch (error) {
      // Internal server error
      return new Response("Internal server error", { status: 500 });
    }
  }

  /**
   * Check if request is a WebSocket upgrade
   */
  private isWebSocketUpgrade(req: Request): boolean {
    const upgrade = req.headers.get("upgrade");
    const connection = req.headers.get("connection");
    return (
      req.method === "GET" &&
      upgrade?.toLowerCase() === "websocket" &&
      connection?.toLowerCase().includes("upgrade") === true
    );
  }

  /**
   * Handle WebSocket connection open
   */
  private handleWebSocketOpen(ws: ServerWebSocket<unknown>): void {
    // Create a new session and handler for this connection
    const session = new Session();
    const handler = new Handler(session, this.handler.rootObject, new NoOpCaller(), [
      this.codec.mimeType(),
    ]);

    // Store session and handler in WebSocket data
    ws.data = { session, handler, codec: this.codec };
  }

  /**
   * Handle WebSocket message
   */
  private async handleWebSocketMessage(
    ws: ServerWebSocket<{ session: Session; handler: Handler; codec: any }>,
    message: string | Buffer,
  ): Promise<void> {
    try {
      const { handler, codec } = ws.data;

      // Convert message to Uint8Array
      const msgBytes =
        typeof message === "string"
          ? new TextEncoder().encode(message)
          : new Uint8Array(message);

      // Decode messages
      let msgSet;
      try {
        msgSet = codec.unmarshalMessages(msgBytes);
      } catch (error) {
        // Parse error - send error response
        const err = parseError(error instanceof Error ? error.message : String(error));
        const errorResp = {
          jsonrpc: "3.0",
          error: err.toObject(),
          id: null,
        };
        const respBytes = codec.marshal(errorResp);
        ws.send(respBytes);
        return;
      }

      // Handle based on batch or single
      let responseMsgSet;
      if (msgSet.isBatch) {
        // Batch request
        const batch = toBatch(msgSet);
        const batchResp = await handler.handleBatch(batch);
        responseMsgSet = batchResponseToMessageSet(batchResp);
      } else {
        // Single request
        const request = toRequest(msgSet);
        const response = await handler.handleRequest(request);
        if (response) {
          responseMsgSet = {
            messages: [response],
            isBatch: false,
          };
        } else {
          // Notification - no response
          return;
        }
      }

      // Encode and send response
      const respBytes = codec.marshalMessages(responseMsgSet);
      ws.send(respBytes);
    } catch (error) {
      console.error("WebSocket message handling error:", error);
    }
  }

  /**
   * Handle WebSocket connection close
   */
  private handleWebSocketClose(
    ws: ServerWebSocket<{ session: Session; handler: Handler; codec: any }>,
  ): void {
    // Dispose all references in the session
    if (ws.data?.session) {
      ws.data.session.disposeAll();
    }
  }
}
